File: bench/Reference/channel_pipeline.py
"""Pipeline benchmark — Python queue.Queue."""

import queue
import threading

CAP = 32
N = 5000


def producer(q):
    for n in range(1, N + 1):
        q.put(n * 2)
    q.put(None)


def main():
    q = queue.Queue(CAP)
    t = threading.Thread(target=producer, args=(q,))
    t.start()
    s = 0
    while True:
        v = q.get()
        if v is None:
            break
        s += v
    t.join()
    print(s)


if __name__ == "__main__":
    main()

File: bench/Reference/channel_pipeline.java
import java.util.concurrent.*;

public class channel_pipeline {
    static final Integer SENTINEL = Integer.MIN_VALUE;
    public static void main(String[] args) throws Exception {
        BlockingQueue<Integer> ch = new ArrayBlockingQueue<>(32);
        Thread producer = new Thread(() -> {
            try {
                for (int n = 1; n <= 5000; n++) ch.put(n * 2);
                ch.put(SENTINEL);
            } catch (InterruptedException ignored) {}
        });
        producer.start();
        long sum = 0;
        Integer v;
        while (!(v = ch.take()).equals(SENTINEL)) sum += v;
        producer.join();
        System.out.println(sum);
    }
}
